CRFModuleValidation/CRFHeartRateProcessor.m

```diff
@@ -33,80 +33,125 @@
+const int CRFHeartRateResolutionHeight = 144;   // lowest resolution on an iPhone 6
 
-@implementation CRFHeartRateProcessor
+@implementation CRFHeartRateProcessor {
+    dispatch_queue_t _processingQueue;
+    NSMutableArray <NSNumber *> * _dataPointsHue;
+    
+    MovieRecorderStatus _status;
+    AVAssetWriter *_assetWriter;
+    BOOL _haveStartedSession;
+    AVAssetWriterInput *_videoInput;
+    
+    __weak id<CRFHeartRateProcessorDelegate> _delegate;
+    dispatch_queue_t _delegateCallbackQueue;
+}
 
-- (instancetype)init {
+- (instancetype)initWithDelegate:(id<CRFHeartRateProcessorDelegate>)delegate callbackQueue:(dispatch_queue_t)queue {
+    NSParameterAssert(delegate != nil);
+    NSParameterAssert(queue != nil);
+    
     self = [super init];
     if (self) {
+        _delegate = delegate;
+        _delegateCallbackQueue = queue;
         _dataPointsHue = [NSMutableArray new];
+        _processingQueue = dispatch_queue_create("org.sagebase.CRF.heartRateSample.processing", DISPATCH_QUEUE_SERIAL);
     }
     return self;
 }
 
-- (void)addDataPoint:(double)hue {
-    if (hue > 0) {
-        // Since the hue for blood is in the red zone which cross the degrees point,
-        // offset that value by 180.
-        double offsetHue = hue + 180.0;
-        if (offsetHue > 360.0) {
-            offsetHue -= 360.0;
-        }
-        [_dataPointsHue addObject:@(offsetHue)];
-    } else {
-        [_dataPointsHue removeAllObjects];
+- (void)appendVideoSampleBuffer:(CMSampleBufferRef)sampleBuffer {
+    if (sampleBuffer == NULL) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"NULL sample buffer" userInfo:nil];
+        return;
     }
-}
-
-- (NSInteger)calculateBPM {
     
-    // If a valid heart rate cannot be calculated then return -1 as an invalid marker
-    if (_dataPointsHue.count < CRFHeartRateMinFrameCount) {
-        return -1;
-    }
-    
-    // Get a window of data points that is the length of the window we are looking at
-    NSUInteger len = CRFHeartRateWindowSeconds * CRFHeartRateFramesPerSecond;
-    NSArray *dataPoints = [_dataPointsHue subarrayWithRange:NSMakeRange(_dataPointsHue.count - len, len)];
-    
-    // If we have enough data points then remove from beginning
-    if (_dataPointsHue.count > CRFHeartRateMinFrameCount) {
-        NSInteger len = _dataPointsHue.count - CRFHeartRateMinFrameCount;
-        [_dataPointsHue removeObjectsInRange:NSMakeRange(0, len)];
-    }
-    
-    // If the heart rate calculated is too low, then it isn't valid
-    NSInteger heartRate = [self calculateBPMWithDataPoints:dataPoints];
-    return heartRate >= 40 ? heartRate : -1;
+    CFRetain(sampleBuffer);
+    dispatch_async(_processingQueue, ^{
+        @autoreleasepool {
+            
+            // Process the sample
+            [self processSampleBuffer:sampleBuffer];
+            
+            // Now, look to see if this sample should be saved to video
+            @synchronized(self) {
```

Comment:
So, @synchronized is fairly heavyweight as thread synch methods go, but no idea whether it's enough to affect collecting samples. On the other hand the only place I see the _status being updated outside of this same serial queue is in a completion handler inside of stopRecording(), and it seems like you could just dispatch that completion handler into this same serial queue asynchronously. I could easily be missing something important and am only mentioning this on the off chance you didn't already think of/try this and rule it out.

Reply:
Nope, intentionally left in all the heavy handed code adapted from the example code from Apple. My initial efforts at tidy, easy to read, obj-c code (gleened from assorted googling) was crashing due to retain loops and deadlocks. Decided that the belt and suspenders was quicker to implement.